### PR TITLE
Conditionally targets NET 9 

### LINKED ISSUE #2551: [Feature Request] conditionally target .NET9-preview
see this .NET8-preview PR as an example of what to do

https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2108
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2121
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2122

Make sure the .NET9-preview is added to the GitHub action and the right builds. We should also have a build variable for .NET 9

start with 9.0.100-preview.2.24157.14
https://dotnet.microsoft.com/en-us/download/dotnet/9.0 and we'll update it every month. Using a * does not work well in this case.

### PR DESCRIPTION
**Pull request by FuPingFranco**
## Description

Conditionally adds NET 9 as a supported target framework for IdentityModel. To test it locally you'll need to have .NET 9 preview SDK installed in your machine then, run "dotnet test Wilson.sln -p:TargetNet9=True" on a shell. VS preview doesn't support NET9 at this moment.

Passing builds:

TargetNet9 = True (https://identitydivision.visualstudio.com/IDDP/_build/results?buildId=1287141&view=logs&j=ca395085-040a-526b-2ce8-bdc85f692774)

TargetNet9 = False (https://identitydivision.visualstudio.com/IDDP/_build/results?buildId=1287142&view=logs&j=ca395085-040a-526b-2ce8-bdc85f692774)

Fixes #2551

### CODE REVIEW

File: test/System.IdentityModel.Tokens.Jwt.Tests/JwtPayloadTest.cs [hunk 1]
```diff
@@ -58,9 +58,13 @@ public void GetSets()
             JwtPayload jwtPayload = new JwtPayload();
             Type type = typeof(JwtPayload);
             PropertyInfo[] properties = type.GetProperties();
+#if NET9_0_OR_GREATER
```
Comment (FuPingFranco) [inline test/System.IdentityModel.Tokens.Jwt.Tests/JwtPayloadTest.cs:61]: This is a temporary fix to complete this work. Need to look in-depth into fixing this test properly on a separate PR.

**Review by JoshLozensky — APPROVED** [review]

File: build/common.props [hunk 2]
```diff
@@ -25,7 +25,7 @@
     <IncludeSymbols>true</IncludeSymbols>
     <SymbolPackageFormat>snupkg</SymbolPackageFormat>
     <TargetFrameworks Condition=" '$(OS)' == 'Windows_NT' ">$(SrcTargets)</TargetFrameworks>
-    <TargetFrameworks Condition=" '$(OS)' != 'Windows_NT' ">$(SrcStandardTargets)</TargetFrameworks>
+    <TargetFramework Condition=" '$(OS)' != 'Windows_NT' ">$(SrcStandardTargets)</TargetFramework>
```
Comment (jmprieur) [inline build/common.props]: Why this change?

File: build/targets.props [hunk 3]
```diff
@@ -1,7 +1,9 @@
 <Project>
   <PropertyGroup>
-    <SrcTargets Condition="'$(LocalBuild)' != 'True'">net461;net462;net472;netstandard2.0;net6.0;net8.0</SrcTargets>
-    <SrcTargets Condition="'$(LocalBuild)' == 'True'">net461;netstandard2.0;net8.0</SrcTargets>
+    <SrcTargets Condition="'$(LocalBuild)' != 'True' and '$(TargetNet9)' != 'True'">net461;net462;net472;netstandard2.0;net6.0;net8.0</SrcTargets>
+    <SrcTargets Condition="'$(LocalBuild)' != 'True' and '$(TargetNet9)' == 'True'">net461;net462;net472;netstandard2.0;net6.0;net8.0;net9.0</SrcTargets>
+    <SrcTargets Condition="'$(LocalBuild)' == 'True' and '$(TargetNet9)' != 'True'">net461;netstandard2.0;net8.0</SrcTargets>
+    <SrcTargets Condition="'$(LocalBuild)' == 'True' and '$(TargetNet9)' == 'True'">net461;netstandard2.0;net8.0;net9.0</SrcTargets>
     <SrcStandardTargets>netstandard2.0</SrcStandardTargets>
```
Comment (jmprieur) [inline build/targets.props:6]: There would be a better way of doing:

```xml
    <SrcTargets Condition="'$(LocalBuild)' != 'True'">net461;net462;net472;netstandard2.0;net6.0;net8.0</SrcTargets>
    <SrcTargets Condition="'$(LocalBuild)' == 'True'">net461;netstandard2.0;net8.0</SrcTargets>
    <SrcTargets Condition="'$(TargetNet9)' == 'True'">$(SrcTargets);net9.0</SrcTargets>
```

File: build/targetsTest.props [hunk 4]
```diff
@@ -1,7 +1,9 @@
 <Project>
   <PropertyGroup>
-    <TestTargets Condition="'$(LocalBuild)' != 'True'">net461;net462;net472;netcoreapp2.1;net6.0;net8.0</TestTargets>
-    <TestTargets Condition="'$(LocalBuild)' == 'True'">net461;netcoreapp2.1;net8.0</TestTargets>
+    <TestTargets Condition="'$(LocalBuild)' != 'True' and '$(TargetNet9)' != 'True'">net461;net462;net472;netcoreapp2.1;net6.0;net8.0</TestTargets>
+    <TestTargets Condition="'$(LocalBuild)' != 'True' and '$(TargetNet9)' == 'True'">net461;net462;net472;netcoreapp2.1;net6.0;net8.0;net9.0</TestTargets>
+    <TestTargets Condition="'$(LocalBuild)' == 'True' and '$(TargetNet9)' != 'True'">net461;netcoreapp2.1;net8.0</TestTargets>
+    <TestTargets Condition="'$(LocalBuild)' == 'True' and '$(TargetNet9)' == 'True'">net461;netcoreapp2.1;net8.0;net9.0</TestTargets>
     <TestOnlyCoreTargets>netcoreapp2.1</TestOnlyCoreTargets>
```
Comment (jmprieur) [inline build/targetsTest.props:6]: Same things:

```xml
   <TestTargets Condition="'$(LocalBuild)' != 'True'">net461;net462;net472;netcoreapp2.1;net6.0;net8.0</TestTargets>
   <TestTargets Condition="'$(LocalBuild)' == 'True'">net461;netcoreapp2.1;net8.0</TestTargets>
   <TestTargets Condition="'$(TargetNet9)' == 'True'">$(TestTargets); net9.0</TestTargets>
```

File: src/Microsoft.IdentityModel.KeyVaultExtensions/Microsoft.IdentityModel.KeyVaultExtensions.csproj [hunk 5]
```diff
@@ -8,6 +8,7 @@
     <GenerateDocumentationFile>true</GenerateDocumentationFile>
     <PackageId>Microsoft.IdentityModel.KeyVaultExtensions</PackageId>
     <TargetFrameworks>netstandard2.0;net6.0;net8.0</TargetFrameworks>
+    <TargetFrameworks Condition="'$(TargetNet9)' == 'True'">$(TargetFrameworks);net9.0</TargetFrameworks>
```
Comment (jmprieur) [inline src/Microsoft.IdentityModel.KeyVaultExtensions/Microsoft.IdentityModel.KeyVaultExtensions.csproj:11]: right. This is the pattern to use.

**Review by jmprieur — CHANGES_REQUESTED** [review]
@FuPingFranco 
There is one line which I think is wrong (hence blocking until you clarify)

Also a few suggestions.

File: build/common.props (hunk 2 above)
Comment (FuPingFranco) [reply build/common.props]: SrcStandardTargets only targets netstandard2.0 in the targets.props. Is it still necessary to use TargetFrameworks for a single target?

Comment (jmprieur) [reply build/common.props]: Good question, @FuPingFranco, and the answer is yes. Please keep TargetFrameworks.
When both as used, this is very confusing to debug (experience from IdWeb)

File: build/targets.props (hunk 3 above)
Comment (FuPingFranco) [reply build/targets.props:6]: Thanks for catching this @jmprieur. This was the initial change I made when I started the work and forgot to clean it before publishing the PR.

File: build/targetsTest.props (hunk 4 above)
Comment (FuPingFranco) [reply build/targetsTest.props:6]: Same as above. Thanks Jean-Marc!

File: build/common.props (hunk 2 above)
Comment (FuPingFranco) [reply build/common.props]: I see, reverting back this change. Thanks Jean-Marc!

**Review by jmprieur — APPROVED** [review]
Nice and clear!
Thanks @FuPingFranco
